compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12647,6 +12659,195 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
+            unsignedMax = isUnsigned  ? TR::getMaxUnsigned<TR::Int16>() : 0;
+            unsignedMin = isUnsigned  ? TR::getMinUnsigned<TR::Int16>() : 0;
+            }
+         else if (firstChild->getOpCode().is4Byte())
+            {
+            signedMax   = !isUnsigned ? TR::getMaxSigned<TR::Int32>()   : 0;
+            signedMin   = !isUnsigned ? TR::getMinSigned<TR::Int32>()   : 0;
+            unsignedMax = isUnsigned  ? TR::getMaxUnsigned<TR::Int32>() : 0;
+            unsignedMin = isUnsigned  ? TR::getMinUnsigned<TR::Int32>() : 0;
+            }
+         else
+            {
+            TR_ASSERT(0, "Unable to handle a number that's bigger than 4 bytes.");
+            }
+
+         if ((isUnsigned && (const3 > unsignedMax || const3 < unsignedMin)) ||
+                 (!isUnsigned && (const3 > signedMax || const3 < signedMin)))
+            {
+            return node;
+            }
+         }
+      else  // Long
+         {
+         if (isUnsigned)
+            {
+            uint64_t uconst1 = firstChild->getSecondChild()->getUnsignedLongInt();
+            uint64_t uconst2 = secondChild->getUnsignedLongInt();
+            uint64_t uconst3 = 0;
+
+            if (isAdd && !IS_UINT64_SUB_UNDERFLOW(const2, const1))
+               {
+               uconst3 = uconst2 - uconst1;
+               }
+            else if (isSub && !IS_UINT64_ADD_OVERFLOW(const2, const1))
+               {
+               uconst3 = uconst2 + uconst1;
+               }
+            else
+               {
+               return node;
+               }
+            }
+         else
+            {
+            const1 = firstChild->getSecondChild()->getConstValue();
+            const2 = secondChild->getConstValue();
+
+            if (isAdd &&
+                    !IS_INT64_SUB_OVERFLOW(const2, const1) &&
+                    !IS_INT64_SUB_UNDERFLOW(const2, const1))
+               {
+               const3 = const2 - const1;
+               }
+            else if (isSub &&
+                     !IS_INT64_ADD_OVERFLOW(const2, const1) &&
+                     !IS_INT64_ADD_UNDERFLOW(const2, const1))
+               {
+               const3 = const2 + const1;
+               }
+            else
+               {
+               return node;
+               }
+            }
+         }
+
+      // Done checking constant values. Change the nodes.
+      dumpOptDetails(s->comp(), "%sEliminating add/sub operation under comparison node n%dn\n",
+                     s->optDetailString(),
+                     node->getGlobalIndex());
+
+      // Handle the iconst node in the compare
+      bool isUnsignedLong = isUnsigned && !isInt;
+      if (secondChild->isSingleRef())
+         {
+         if (isUnsignedLong)
+            {
+            secondChild->setUnsignedLongInt(const3);
+            }
+         else
+            {
+            secondChild->setConstValue(const3);
+            }
+         }
```

Comment:
A node shouldn't be transmuted in a way that changes its value, so this should always create a new constant